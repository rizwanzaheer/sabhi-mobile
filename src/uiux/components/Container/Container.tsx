import * as React from 'react'
import { View, ViewStyle, StyleSheet } from 'react-native'
import { withTheme } from '../../theming'
import * as Kancha from '../../types'

interface ContainerProps {
  /** Test ID used for e2e tests */
  testID?: string

  /** Width */
  w?: string | number | undefined

  /** Height */
  h?: string | number | undefined

  /** Bottom */
  b?: string | number | undefined

  /** Right */
  r?: string | number | undefined

  /** Border radius */
  br?: number | undefined

  /** Flex */
  flex?: number | undefined

  /** Pre-defined backgrounds accordign to the theme. use these where possible. */
  background?: Kancha.BrandPropOptions

  /** Temporary option to create custom color. Avoid is possible and deprecate if you can by modifying the theme */
  backgroundColor?: string

  /** Flex direction */
  flexDirection?: 'row' | 'column' | 'row-reverse' | 'column-reverse' | undefined

  /** Align items */
  alignItems?: 'flex-start' | 'flex-end' | 'center' | 'stretch' | 'baseline' | undefined

  /** Justify Content */
  justifyContent?:
    | 'flex-start'
    | 'flex-end'
    | 'center'
    | 'space-between'
    | 'space-around'
    | 'space-evenly'
    | undefined

  /** Set the bottom divider */
  dividerBottom?: boolean

  /** et the top divider */
  dividerTop?: boolean

  /** Set the bottom margin */
  margin?: number | boolean | undefined

  /** Set the bottom margin */
  marginBottom?: number | boolean | undefined

  /** Set the top margin */
  marginTop?: number | boolean | undefined

  /** Set the bottom margin */
  marginLeft?: number | boolean | undefined

  /** Set the top margin */
  marginRight?: number | boolean | undefined

  /** Set the default padding */
  padding?: number | boolean | undefined

  /** Set the default paddingHorizontal */
  paddingHorizontal?: number | boolean | undefined

  /** Set the bottom padding */
  paddingBottom?: number | boolean | undefined

  /** Set the top padding */
  paddingTop?: number | boolean | undefined

  /** Set the left padding */
  paddingLeft?: number | boolean | undefined

  /** Set the right padding */
  paddingRight?: number | boolean | undefined

  /** Enable border for debugging layouts */
  debugBorder?: boolean

  /** Enable border for debugging layouts */
  borderColor?: string

  /** Enable border for debugging layouts */
  borderWidth?: number

  /** Change debug border color */
  debugBorderColor?: string | undefined

  /** Add addionaly custom styles for a container. Use sparingly!! */
  viewStyle?: ViewStyle

  /** Disable the view from being able to detect interactions */
  disabled?: boolean

  /** A shadow level to apply */
  shadow?: number

  /** An opacity level. Use 1, 2, 3 etc  */
  opacity?: number

  theme: any
}

/**
 * `Container` is the most fundamental building block in Kancha and should be used anywhere you would use a `View`.
 * In most cases the built-in React Native `View` component is used in conjunction with some some simple layout styles like flexbox casuing a mess of styles.
 * Using a `Container` these can just be added as props for declarative layouts that are easy to read.
 *
 * ```tsx
 * <Container br={4} flex={1} alignItems={'center'}>
 *   <Text>Hey, This is a container</Text>
 * </Container>
 * ```
 */
const Container: React.FunctionComponent<ContainerProps> = props => {
  const DividerBottomStyles: ViewStyle = {
    borderBottomWidth: StyleSheet.hairlineWidth,
    borderBottomColor: props.theme.colors.primary.divider,
  }

  const DividerTopStyles: ViewStyle = {
    borderTopWidth: StyleSheet.hairlineWidth,
    borderTopColor: props.theme.colors.primary.divider,
  }

  const BaseStyles: ViewStyle = {
    /** Basic view styles */
    width: props.w,
    height: props.h,
    flex: props.flex,
    flexDirection: props.flexDirection,
    alignItems: props.alignItems,
    justifyContent: props.justifyContent,
    backgroundColor: props.background && props.theme.colors[props.background].background,
    borderRadius: props.br,
  }

  /** Conditionally spread props down to the View as styles */
  const styles: ViewStyle = {
    ...BaseStyles,
    ...(props.dividerBottom ? DividerBottomStyles : {}),
    ...(props.dividerTop ? DividerTopStyles : {}),
    ...(props.borderColor ? { borderColor: props.borderColor } : {}),
    ...(props.borderWidth ? { borderWidth: props.borderWidth } : {}),
    ...(props.backgroundColor ? { backgroundColor: props.backgroundColor } : {}),
    ...(props.debugBorder ? { borderWidth: 1, borderColor: 'red' } : {}),
    ...(props.b !== undefined ? { position: 'absolute', bottom: props.b, width: '100%' } : {}),
    ...(props.r !== undefined ? { position: 'absolute', bottom: props.r, width: '100%' } : {}),
    ...(props.disabled ? { opacity: 0.5 } : {}),
    ...(props.shadow
      ? {
          elevation: props.shadow * 10,
          shadowColor: '#ffffff',
          shadowOpacity: 0.2,
          shadowRadius: props.shadow * 5,
          backgroundColor: props.backgroundColor
            ? props.background && props.theme.colors[props.background].background
            : '#ffffff',
        }
      : {}),

    /** Margins */
    margin: typeof props.margin === 'boolean' ? props.theme.spacing.default : props.margin,
    marginBottom: typeof props.marginBottom === 'boolean' ? props.theme.spacing.default : props.marginBottom,
    marginTop: typeof props.marginTop === 'boolean' ? props.theme.spacing.default : props.marginTop,
    marginLeft: typeof props.marginLeft === 'boolean' ? props.theme.spacing.default : props.marginLeft,
    marginRight: typeof props.marginRight === 'boolean' ? props.theme.spacing.default : props.marginRight,

    /** Paddings */
    padding: typeof props.padding === 'boolean' ? props.theme.spacing.default : props.padding,
    paddingHorizontal:
      typeof props.paddingHorizontal === 'boolean' ? props.theme.spacing.default : props.paddingHorizontal,

    paddingBottom:
      typeof props.paddingBottom === 'boolean' ? props.theme.spacing.default : props.paddingBottom,
    paddingTop: typeof props.paddingTop === 'boolean' ? props.theme.spacing.default : props.paddingTop,
    paddingLeft: typeof props.paddingLeft === 'boolean' ? props.theme.spacing.default : props.paddingLeft,
    paddingRight: typeof props.paddingRight === 'boolean' ? props.theme.spacing.default : props.paddingRight,

    /** Viewstyle props will overide all options */
    ...(props.viewStyle ? { ...props.viewStyle } : {}),
  }

  return (
    <View testID={props.testID} style={styles} pointerEvents={props.disabled ? 'none' : 'auto'}>
      {props.children}
    </View>
  )
}

export default withTheme(Container)
