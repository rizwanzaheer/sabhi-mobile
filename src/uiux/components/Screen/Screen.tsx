import * as React from 'react'
import { ImageSourcePropType, SafeAreaView, StatusBar, ImageBackground } from 'react-native'
import { KeyboardAwareScrollView } from 'react-native-keyboard-aware-scroll-view'
import Container from '../Container/Container'
import { withTheme } from '../../theming'
import * as Kancha from '../../types'

interface ScreenProps {
  /**
   * Safe area for entire View. This will over-ride other safearea options
   */
  safeArea?: boolean

  /**
   * Background type
   */
  background?: Kancha.BrandPropOptions

  /**
   * Safe background color. This will over-ride other safearea background color options
   */
  safeAreaBackground?: string

  /**
   * Safearea top
   */
  safeAreaTop?: boolean

  /**
   * Safearea top background color
   */
  safeAreaTopBackground?: string

  /**
   * Safearea bottom
   */
  safeAreaBottom?: boolean

  /**
   * Safearea bottom background color
   */
  safeAreaBottomBackground?: string

  /**
   * Enable scrolling view
   */
  scrollEnabled?: boolean

  /**
   * Provide a banner component for the view
   */
  bannerComponent?: React.ReactNode

  /**
   * Set a background image for the view
   */
  backgroundColor?: string

  /**
   * Set a background image for the view
   */
  backgroundImage?: ImageSourcePropType

  /**
   * Hide the status bar
   */
  statusBarHidden?: boolean

  /**
   * Set the statusbar style. Override the theme's default
   */
  statusBarStyle?: string

  /**
   * Provide a footer component
   */
  footerComponent?: React.ReactNode

  /**
   * Show a divider line between footer component and view
   */
  footerDivider?: boolean

  /**
   * Passed in by the HOC. Safe to ignore.
   */
  theme: any

  /**
   * An action button that floats over the content
   */
  fabButton?: React.ReactNode
}

/**
 * Screen component provides useful UI helpers that are needed on most screens.
 */
const Screen: React.FC<ScreenProps> = props => {
  const mainContent = (
    <React.Fragment>
      <StatusBar
        hidden={props.statusBarHidden}
        barStyle={props.statusBarStyle ? props.statusBarStyle : props.theme.statusBarStyle}
        animated={true}
        showHideTransition={'slide'}
      />
      <Container flex={1} background={props.background}>
        {props.children}
      </Container>
    </React.Fragment>
  )
  const scrollViewContent = (
    <KeyboardAwareScrollView
      style={{
        backgroundColor: props.background
          ? props.theme.colors[props.background].background
          : props.backgroundColor
          ? props.backgroundColor
          : props.theme.colors.secondary.background,
      }}
      contentInsetAdjustmentBehavior={'never'}
    >
      {mainContent}
    </KeyboardAwareScrollView>
  )
  const scrollSafeViews = (
    <React.Fragment>
      {(props.safeArea || props.safeAreaTop) && (
        <SafeAreaView style={{ backgroundColor: props.safeAreaBackground || props.safeAreaTopBackground }} />
      )}
      {props.scrollEnabled ? scrollViewContent : mainContent}
      {props.footerComponent && (
        <Container
          paddingTop={true}
          dividerTop={props.footerDivider}
          backgroundColor={props.theme.colors.primary.background}
        >
          {props.footerComponent}
        </Container>
      )}
      {props.fabButton && (
        <Container viewStyle={{ position: 'absolute', width: '100%', bottom: 40, alignItems: 'center' }}>
          {props.fabButton}
        </Container>
      )}
      {(props.safeArea || props.safeAreaBottom) && (
        <SafeAreaView
          style={{
            backgroundColor:
              props.safeAreaBackground ||
              props.safeAreaBottomBackground ||
              props.theme.colors.primary.background,
          }}
        />
      )}
    </React.Fragment>
  )
  const backgroundImage = (
    <ImageBackground
      style={{ flex: 1 }}
      source={props.backgroundImage ? props.backgroundImage : {}}
      resizeMode={'cover'}
    >
      {scrollSafeViews}
    </ImageBackground>
  )

  return props.backgroundImage ? backgroundImage : scrollSafeViews
}

export default withTheme(Screen)
