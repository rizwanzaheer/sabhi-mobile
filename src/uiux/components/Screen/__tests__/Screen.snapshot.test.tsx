import * as React from 'react'
import { render } from 'react-native-testing-library'
import Screen from '../Screen'
import Container from '../../Container/Container'

describe('Component(snapshots): Screen', () => {
  it('should render with no props', () => {
    const tree = render(<Screen />)

    expect(tree).toMatchSnapshot()
  })

  it('should render with props', () => {
    const tree = render(
      <Screen
        background={'primary'}
        safeArea={true}
        safeAreaBackground="white"
        scrollEnabled={true}
        statusBarStyle={'light-content'}
        footerComponent={<Container />}
      />,
    )

    expect(tree).toMatchSnapshot()
  })

  it('should render with a custom background color', () => {
    const tree = render(<Screen backgroundColor={'#000000'} />)

    expect(tree).toMatchSnapshot()
  })

  it('should render with no statusbar and independant safe areas', () => {
    const tree = render(
      <Screen
        statusBarHidden={true}
        safeAreaTop={true}
        safeAreaTopBackground="white"
        safeAreaBottom={true}
        safeAreaBottomBackground="white"
        backgroundImage={{ uri: 'https://theinternet.com/image.jpg' }}
      />,
    )

    expect(tree).toMatchSnapshot()
  })
})
