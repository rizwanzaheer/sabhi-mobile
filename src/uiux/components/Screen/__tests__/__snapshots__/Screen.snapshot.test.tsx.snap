// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component(snapshots): Screen should render with a custom background color 1`] = `
<View
  pointerEvents="auto"
  style={
    Object {
      "alignItems": undefined,
      "backgroundColor": undefined,
      "borderRadius": undefined,
      "flex": 1,
      "flexDirection": undefined,
      "height": undefined,
      "justifyContent": undefined,
      "margin": undefined,
      "marginBottom": undefined,
      "marginLeft": undefined,
      "marginRight": undefined,
      "marginTop": undefined,
      "padding": undefined,
      "paddingBottom": undefined,
      "paddingHorizontal": undefined,
      "paddingLeft": undefined,
      "paddingRight": undefined,
      "paddingTop": undefined,
      "width": undefined,
    }
  }
/>
`;

exports[`Component(snapshots): Screen should render with no props 1`] = `
<View
  pointerEvents="auto"
  style={
    Object {
      "alignItems": undefined,
      "backgroundColor": undefined,
      "borderRadius": undefined,
      "flex": 1,
      "flexDirection": undefined,
      "height": undefined,
      "justifyContent": undefined,
      "margin": undefined,
      "marginBottom": undefined,
      "marginLeft": undefined,
      "marginRight": undefined,
      "marginTop": undefined,
      "padding": undefined,
      "paddingBottom": undefined,
      "paddingHorizontal": undefined,
      "paddingLeft": undefined,
      "paddingRight": undefined,
      "paddingTop": undefined,
      "width": undefined,
    }
  }
/>
`;

exports[`Component(snapshots): Screen should render with no statusbar and independant safe areas 1`] = `
<View
  accessibilityIgnoresInvertColors={true}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Image
    resizeMode="cover"
    source={
      Object {
        "uri": "https://theinternet.com/image.jpg",
      }
    }
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "height": undefined,
          "width": undefined,
        },
        undefined,
      ]
    }
  />
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Object {
        "backgroundColor": "white",
      }
    }
  />
  <View
    pointerEvents="auto"
    style={
      Object {
        "alignItems": undefined,
        "backgroundColor": undefined,
        "borderRadius": undefined,
        "flex": 1,
        "flexDirection": undefined,
        "height": undefined,
        "justifyContent": undefined,
        "margin": undefined,
        "marginBottom": undefined,
        "marginLeft": undefined,
        "marginRight": undefined,
        "marginTop": undefined,
        "padding": undefined,
        "paddingBottom": undefined,
        "paddingHorizontal": undefined,
        "paddingLeft": undefined,
        "paddingRight": undefined,
        "paddingTop": undefined,
        "width": undefined,
      }
    }
  />
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Object {
        "backgroundColor": "white",
      }
    }
  />
</View>
`;

exports[`Component(snapshots): Screen should render with props 1`] = `
Array [
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Object {
        "backgroundColor": "white",
      }
    }
  />,
  <RCTScrollView
    automaticallyAdjustContentInsets={false}
    contentInset={
      Object {
        "bottom": 0,
      }
    }
    contentInsetAdjustmentBehavior="never"
    enableAutomaticScroll={true}
    enableOnAndroid={false}
    enableResetScrollToCoords={true}
    extraHeight={75}
    extraScrollHeight={0}
    getScrollResponder={[Function]}
    handleOnScroll={[Function]}
    keyboardDismissMode="interactive"
    keyboardOpeningTime={250}
    keyboardSpace={0}
    onScroll={[Function]}
    resetKeyboardSpace={[Function]}
    scrollEventThrottle={1}
    scrollForExtraHeightOnAndroid={[Function]}
    scrollIntoView={[Function]}
    scrollToEnd={[Function]}
    scrollToFocusedInput={[Function]}
    scrollToPosition={[Function]}
    showsVerticalScrollIndicator={true}
    style={
      Object {
        "backgroundColor": "#FFFFFF",
      }
    }
    update={[Function]}
    viewIsInsideTabBar={false}
  >
    <View>
      <View
        pointerEvents="auto"
        style={
          Object {
            "alignItems": undefined,
            "backgroundColor": "#FFFFFF",
            "borderRadius": undefined,
            "flex": 1,
            "flexDirection": undefined,
            "height": undefined,
            "justifyContent": undefined,
            "margin": undefined,
            "marginBottom": undefined,
            "marginLeft": undefined,
            "marginRight": undefined,
            "marginTop": undefined,
            "padding": undefined,
            "paddingBottom": undefined,
            "paddingHorizontal": undefined,
            "paddingLeft": undefined,
            "paddingRight": undefined,
            "paddingTop": undefined,
            "width": undefined,
          }
        }
      />
    </View>
  </RCTScrollView>,
  <View
    pointerEvents="auto"
    style={
      Object {
        "alignItems": undefined,
        "backgroundColor": "#FFFFFF",
        "borderRadius": undefined,
        "flex": undefined,
        "flexDirection": undefined,
        "height": undefined,
        "justifyContent": undefined,
        "margin": undefined,
        "marginBottom": undefined,
        "marginLeft": undefined,
        "marginRight": undefined,
        "marginTop": undefined,
        "padding": undefined,
        "paddingBottom": undefined,
        "paddingHorizontal": undefined,
        "paddingLeft": undefined,
        "paddingRight": undefined,
        "paddingTop": 15,
        "width": undefined,
      }
    }
  >
    <View
      pointerEvents="auto"
      style={
        Object {
          "alignItems": undefined,
          "backgroundColor": undefined,
          "borderRadius": undefined,
          "flex": undefined,
          "flexDirection": undefined,
          "height": undefined,
          "justifyContent": undefined,
          "margin": undefined,
          "marginBottom": undefined,
          "marginLeft": undefined,
          "marginRight": undefined,
          "marginTop": undefined,
          "padding": undefined,
          "paddingBottom": undefined,
          "paddingHorizontal": undefined,
          "paddingLeft": undefined,
          "paddingRight": undefined,
          "paddingTop": undefined,
          "width": undefined,
        }
      }
    />
  </View>,
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Object {
        "backgroundColor": "white",
      }
    }
  />,
]
`;
